src/vehicle_autodrive.cpp

```diff
@@ -0,0 +1,960 @@
+    node_address rot_addr = addr;
+    rot_addr.x -= NAV_MAP_SIZE / 2;
+    rot_addr.y -= NAV_MAP_SIZE / 2;
+    if( data.goal_zone.p_min.y == 0 || data.goal_zone.p_max.y == NAV_MAP_SIZE ) {
+        rot_addr.facing_dir += NUM_ORIENTATIONS / 4;
+        std::swap( rot_addr.x, rot_addr.y );
+        rot_addr.x *= -1;
+    }
+    if( data.goal_zone.p_min.x == 0 || data.goal_zone.p_max.y == NAV_MAP_SIZE ) {
+        rot_addr.facing_dir += NUM_ORIENTATIONS / 2;
+        rot_addr.x *= -1;
+        rot_addr.y *= -1;
+    }
+    rot_addr.x += NAV_MAP_SIZE / 2;
+    rot_addr.y += NAV_MAP_SIZE / 2;
+    rot_addr.facing_dir = normalize_orientation( rot_addr.facing_dir );
+    const int dx = NAV_MAP_SIZE - 1 - rot_addr.x;
+    const int dy = NAV_MAP_SIZE / 2 - rot_addr.y;
+    const orientation goal_dir = approx_orientation( dx, dy );
+    ret.score += forward_dist_mult * dx;
+    ret.score += side_dist_mult * std::abs( dy );
+    ret.score += dx > NAV_MAP_SIZE / 3 ? angle_mult * std::abs( orientation_diff( rot_addr.facing_dir,
+                 goal_dir ) ) : 0;
+    return ret;
+}
+
+
+void vehicle::autodrive_controller::compute_next_nodes( const node_address &addr,
+        const navigation_node &node, int target_speed_tps,
+        std::vector<std::pair<node_address, navigation_node>> &next_nodes )
+const
+{
+    const int sign = target_speed_tps > 0 ? 1 : -1;
+    const int target_speed = target_speed_tps * VMIPH_PER_TPS;
+    int next_speed = target_speed;
+    int num_tiles_to_move = std::abs( target_speed_tps );
+    if( target_speed_tps > 1 && node.speed < target_speed ) {
+        const int cur_tps = std::min( std::max( node.speed / VMIPH_PER_TPS, 0 ), data.max_speed_tps - 1 );
+        next_speed = std::min( std::max<int>( node.speed, 0 ) + data.acceleration[cur_tps], target_speed );
+        num_tiles_to_move = next_speed / VMIPH_PER_TPS;
+    }
+    for( int steer = -data.max_steer; steer <= data.max_steer; steer++ ) {
+        node_address next_addr = addr;
+        next_addr.facing_dir = normalize_orientation( addr.facing_dir + steer );
+        tileray tdir = data.profiles[next_addr.facing_dir].tdir;
+        if( steer == 0 ) {
+            tdir.advance( node.tileray_steps );
+        }
+        bool ok = true;
+        for( int i = 0; i < num_tiles_to_move; i++ ) {
+            tdir.advance( sign );
+            next_addr.x += tdir.dx();
+            next_addr.y += tdir.dy();
+            if( !data.nav_zone.contains( {next_addr.x, next_addr.y} ) ||
+                !data.valid_position[next_addr.facing_dir][next_addr.x][next_addr.y] ) {
+                ok = false;
+                break;
+            }
+        }
+        if( !ok ) {
+            continue;
+        }
+        navigation_node next_node;
+        next_node.prev = addr;
+        next_node.tileray_steps = ( steer == 0 ) ? node.tileray_steps + 1 : 1;
+        // TODO: improve this
+        next_node.cost = node.cost + 10 + ( steer ? 1 : 0 );
```

Comment:
magic numbers are bad.  I mean, I can barely understand this code as it is, I can't figure out if this 10 is the same meaning as the tileray advance 10 or not.

Reply:
added named constants